lib/spack/spack/util/url.py

```diff
@@ -0,0 +1,230 @@
+            # not related to whether Spack validates certificates for https
+            # connections.  That is controlled by Spack's config and/or
+            # -k/--insecure CLI flags.
+            if netloc.endswith(':') and path.startswith('//'):
+                netloc_host, path = (path[2:].split('/', 1) + [''])[:2]
+                netloc = '//'.join((netloc, netloc_host))
+
+            path_tokens = [part for part in _split_all(path)
+                    if part and part != '/']
+
+            extra_attrs["s3_bucket"] = (
+                    path_tokens.pop(0) if path_tokens else None)
+            path = pathjoin('', *path_tokens)
+
+            credentials, netloc = (netloc.split('@', 1) + [None])[:2]
+            if netloc is None:
+                netloc, credentials = credentials, netloc
+
+            if credentials:
+                key_id, key_secret = (credentials.split(':', 1) + [None])[:2]
+
+                if key_secret is None:
+                    extra_attrs["s3_profile"] = key_id
+                else:
+                    extra_attrs["s3_access_key_id"] = key_id
+                    extra_attrs["s3_secret_access_key"] = key_secret
+
+    result = ParseResultWrapper(scheme=scheme,
+                                netloc=netloc,
+                                path=path,
+                                params=params,
+                                query=query,
+                                fragment=None)
+
+    for key, val in extra_attrs.items():
+        setattr(result, key, val)
+
+    return result
+
+
+def format(parsed_url):
+    if not isinstance(parsed_url, ParseResultWrapper):
+        parsed_url = parse(parsed_url)
+
+    (scheme, netloc, path, params, query, _) = parsed_url
+
+    scheme = scheme.lower()
+
+    if scheme == 's3':
+        path = pathjoin(
+                *[x for x in ('/', parsed_url.s3_bucket, path) if x])
+
+        credentials = None
+
+        try:
+            credentials = parsed_url.s3_profile
+        except AttributeError:
+            pass
+
+        try:
+            if not credentials:
+                credentials = ':'.join((
+                    parsed_url.s3_access_key_id,
+                    parsed_url.s3_secret_access_key))
+        except AttributeError:
+            pass
+
+        if credentials:
+            netloc = '@'.join((credentials, netloc))
+
+    # Workaround a quirk of urlparse where the double-slash after the [scheme]
+    # is left out in the case of no [netloc] and a [scheme] that is not natively
+    # recognized (e.g.: s3).  S3 URLs will usually specify no [netloc], instead
+    # defaulting to AWS and using other mechanisms, like environment variables,
+    # to provide access credentials.
+    #
+    # The point of this workaround is to manipulate these URLs so that instead
+    # of being formatted like so:
+    #
+    #  s3:/my-bucket
+    #
+    # ... they would be (correctly) formatted like so:
+    #
+    #  s3:///my-bucket
```

Comment:
Are s3 URLs with 3 slashes a thing?  I was skimming this and didn't see any:

https://gpdb.docs.pivotal.io/43320/admin_guide/load/topics/g-s3-protocol.html

Also does Boto not provide any utilities to deal with some of the things this function is doing?

Reply:
> @opadron: I haven't had a chance to go over all the questions here, but I added a bunch of review comments. In general, the high level plan looks good, and the refactors are good. However:
> 
> 1. A lot of the mirror code is repetitive and overly complex. I recommend introducing a `Mirror` class to take care of that.  Currently too much of the code deals directly with YAML instead of with a `Mirror` abstraction.

Agreed.

> 2. Boxes are checked on the PR, but I don't see where the commands now have push or pull options.  Where is that implemented?

Fetching and pushing is implemented directly in the mirror and buildcache commands.  With these changes, `s3://` URLs (or mirror names referring to such URLs) can be provided anywhere where these commands expected a directory.  In these cases, the commands work exactly as they did before, except they handle uploading to- and/or downloading from- S3 buckets in a way that is transparent to the user.

For example: `spack buildcache create -d s3:///spack-public/mirror zlib` populates (creating, if necessary) a buildcache with zlib binaries.  It works exactly the same as before, except now it also works with S3 buckets.

> 3. We discussed making `/var/spack/cache` the "default" mirror where build caches would be created and where things would be created if no `--directory` argument is specified for `spack mirror create` and `spack buildcache create`.  Is that planned?
> 4. There is a lot of stuff in here for parsing URLs and for things like `noopify`, and I am not sure how it factors in or why some of it is needed.  Maybe we can talk about that at an upcoming meeting.

Reply:
> Are s3 URLs with 3 slashes a thing? I was skimming this and didn't see any:
> 
> https://gpdb.docs.pivotal.io/43320/admin_guide/load/topics/g-s3-protocol.html
> 
> Also does Boto not provide any utilities to deal with some of the things this function is doing?

To my knowledge there isn't a standard defining what an S3 URL should have or how it should be formatted.  I tried defining a format that could encode all the information we might need to make the connection with what I hoped are reasonable defaults.  The format documented in your link is actually very similar to what I've come up with.  I will update with documentation for the format, which should answer this and possibly other questions you might have about it.

Boto does not provide any handling of S3 URLs because they're not officially a thing.

Reply:
If they're not officially a thing, then we shouldn't invent a new thing.  It looks like S3 URLs are pretty recognizable as they are now, e.g., [from here](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html) it looks like there are three forms:
* http://<bucket>.s3-<aws-region>.amazonaws.com
* http://<bucket>.s3.amazonaws.com
* http://s3-<aws-region>.amazonaws.com/<bucket>

If those are already copy/pastable from somewhere in AWS, I do not think we need a new URL scheme.  We can infer what we need to know about the bucket from the https URL.

Reply:
The "Accessing a Bucket" section in the docs you link fails to cover a lot of important details. 
 They don't really clarify what is meant by "access", only having this relevant piece to say:

> If you access a bucket programmatically, note that Amazon S3 supports RESTful architecture in which your buckets and objects are resources, each with a resource URI that uniquely identifies the resource.

It says nothing about what actual operations are supported or what headers need to be included for authentication, access control, or billing purposes.  This omission also appears deliberate, as all the AWS documentation I've read on the matter prefers to point you to their console UI, CLI, SDKs or boto.

Regardless, even if this documentation did exist (which it must, somewhere), we'd still need to provide extra information with each request (likely through headers) that is not encoded in any of the URL examples you point to, nor in any examples I've seen in the wild, other than those like the ones in the pivotal docs you linked above.

There's no way an http URL, on its own, is ever going to have enough information for Spack to work with the buckets they may refer to.  And the story gets worse once you consider non-AWS endpoints.

I'm not a fan of inventing a new URL scheme, either, but there's only so much we can do to avoid it if we want to work with S3 buckets.  My hope is that we can take some comfort in that

1. There is some prior art (just nothing standardized or in library form) that I used as-is as my starting point, so I'm not really "inventing" all that much, just reimplementing, and
1. The code is already written and works.  The URL scheme just needs to be documented.